File: FacebookInterviewQuestions/maximumSubarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
  if (nums.length === 0) {
    return [];
  }
  let maxEndingHere = nums[0];
  let maxSoFar = nums[0];

  for (let i = 1; i < nums.length; i++) {
    maxEndingHere = Math.max(maxEndingHere + nums[i], nums[i]);
    maxSoFar = Math.max(maxSoFar, maxEndingHere);
  }

  return maxSoFar;
};

File: python/EasyInterviewQuestions/dynamic_programming/maximum_subarray.py
class Solution:
    def maxSubArray(self, nums):

        maxEndingHere = nums[0]
        maxSoFar = nums[0]
        n = len(nums)

        for i in range(1, n):

            maxEndingHere = max(nums[i], maxEndingHere + nums[i])
            maxSoFar = max(maxSoFar, maxEndingHere)

        return maxSoFar

